mlir-sycl/lib/Dialect/IR/SYCLOpsTypes.cpp

```diff
@@ -129,3 +129,40 @@ mlir::sycl::getDerivedTypes(mlir::TypeID TypeID) {
             mlir::sycl::RangeType::getTypeID()};
   return {};
 }
+
+mlir::LogicalResult
+mlir::sycl::VecType::verify(llvm::function_ref<InFlightDiagnostic()> EmitError,
+                            mlir::Type DataT, int NumElements,
+                            llvm::SmallVector<mlir::Type, 4>) {
+  if (!(NumElements == 1 || NumElements == 2 || NumElements == 3 ||
+        NumElements == 4 || NumElements == 8 || NumElements == 16)) {
+    return EmitError() << "SYCL vector types can only hold 1, 2, 3, 4, 8 or 16 "
+                          "elements. Got "
+                       << NumElements << ".";
+  }
+
+  if (!DataT.isa<IntegerType, FloatType>()) {
+    return EmitError()
+           << "SYCL vector types can only hold basic scalar types. Got "
+           << DataT << ".";
+  }
+
+  if (auto IntTy = DataT.dyn_cast<IntegerType>()) {
+    const auto Width = IntTy.getWidth();
+    if (!(Width == 1 || Width == 8 || Width == 16 || Width == 32 ||
+          Width == 64)) {
+      return EmitError() << "Integer SYCL vector element types can only be i1, "
```

Comment:
Is it true that when the element type is a i64 then NumElements cannot be larger than 2? We should check that combinations of element width and number of elements are legal.

Reply:
I cannot find that affirmation [in the SYCL spec](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_vec_interface). Also:
```cpp
extern "C" SYCL_EXTERNAL void cons_12() {
  auto vec = sycl::vec<sycl::cl_long, 8>{};
}
```
works fine for me using `clang++`.

Reply:
Also, the types `sycl::long8` and `sycl::long16` are defined, so I assume these are legal.